dpc/src/record/encoded.rs

```diff
@@ -16,45 +16,90 @@
+
+/// Encode a base field element bytes to a group representation.
+pub fn encode_to_field<F: PrimeField>(x_bytes: &[u8]) -> Result<Vec<F>, DPCError> {
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(x_bytes.len() * 8 + 1);
+    for byte in x_bytes.iter() {
+        let mut byte = byte.clone();
+        for _ in 0..8 {
+            bits.push(byte & 1 == 1);
+            byte >>= 1;
+        }
+    }
+    bits.push(true);
+    // The last bit indicates the end of the actual data, which is used in decoding to
+    // make sure that the length is correct.
 
-use itertools::Itertools;
-use std::marker::PhantomData;
-
-/// Encode a base field element bytes to a group representation
-pub fn encode_to_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    x_bytes: &[u8],
-) -> Result<(<G as ProjectiveCurve>::Affine, bool), DPCError> {
-    // TODO (howardwu): Remove this hardcoded value and use BaseField's size in bits to pad length.
-    let mut bytes = x_bytes.to_vec();
-    while bytes.len() < 32 {
-        bytes.push(0)
+    let mut res = Vec::<F>::with_capacity((bits.len() + capacity - 1) / capacity);
+    for chunk in bits.chunks(capacity) {
+        res.push(F::from_repr(F::BigInteger::from_bits_le(chunk)).unwrap());
     }
 
-    let input = P::BaseField::read_le(&bytes[..])?;
-    let (output, fq_high) = Elligator2::<P, G>::encode(&input)?;
-    Ok((output, fq_high))
+    Ok(res)
 }
 
-/// Decode a group into the byte representation of a base field element
-pub fn decode_from_group<P: MontgomeryParameters + TwistedEdwardsParameters, G: ProjectiveCurve>(
-    affine: <G as ProjectiveCurve>::Affine,
-    fq_high: bool,
-) -> Result<Vec<u8>, DPCError> {
-    let output = Elligator2::<P, G>::decode(&affine, fq_high)?;
-    Ok(to_bytes_le![output]?)
+/// Decode a group into the byte representation of a base field element.
+pub fn decode_from_field<F: PrimeField>(field_elements: &[F]) -> Result<Vec<u8>, DPCError> {
+    if field_elements.len() == 0 {
+        return Err(DPCError::EncodingError(EncodingError::Message(
+            "The encoded record must consist of at least one field element.".to_string(),
+        )));
+    }
+    // There is at least one field element due to the additional true bit.
+
+    let capacity = <F::Parameters as FieldParameters>::CAPACITY as usize;
+
+    let mut bits = Vec::<bool>::with_capacity(field_elements.len() * capacity);
+    for elem in field_elements.iter() {
+        let elem_bits = elem.to_repr().to_bits_le();
+        bits.extend_from_slice(&elem_bits[0..capacity]); // only keep `capacity` bits, discarding the highest bit.
+    }
+
+    // Find the last bit of `true`, which should be within the last `CAPACITY` bits.
+    // If it does not exist, this is a `DPCError::EncodingError`.
+    let mut last_bit_pos = None;
+    for i in ((bits.len() - capacity)..(bits.len())).rev() {
```

Comment:
```suggestion
    for i in (0..bits.len()).rev().take(capacity) {
```

Reply:
@ljedrz looks like we came up with two approaches.

@weikengchen feel free to take your pick.